File: src/com/mistraltech/smogen/codegenerator/matchergenerator/MatcherInterfaceCodeWriter.java
package com.mistraltech.smogen.codegenerator.matchergenerator;

import com.intellij.psi.PsiPackage;
import com.mistraltech.smogen.codegenerator.javabuilder.AnnotationBuilder;
import com.mistraltech.smogen.codegenerator.javabuilder.InterfaceBuilder;
import com.mistraltech.smogen.codegenerator.javabuilder.InterfaceMethodBuilder;
import com.mistraltech.smogen.codegenerator.javabuilder.JavaDocumentBuilder;
import com.mistraltech.smogen.codegenerator.javabuilder.MethodBuilder;
import com.mistraltech.smogen.codegenerator.javabuilder.MethodSignatureBuilder;
import com.mistraltech.smogen.codegenerator.javabuilder.NewInstanceBuilder;
import com.mistraltech.smogen.codegenerator.javabuilder.TypeBuilder;
import com.mistraltech.smogen.codegenerator.javabuilder.TypeParameterDeclBuilder;
import com.mistraltech.smogen.property.Property;
import com.mistraltech.smogen.property.PropertyLocator;
import org.jetbrains.annotations.NotNull;

import java.util.Arrays;
import java.util.List;

import static com.mistraltech.smogen.codegenerator.javabuilder.AnnotationBuilder.anAnnotation;
import static com.mistraltech.smogen.codegenerator.javabuilder.ExpressionBuilder.anExpression;
import static com.mistraltech.smogen.codegenerator.javabuilder.ExpressionTextBuilder.expressionText;
import static com.mistraltech.smogen.codegenerator.javabuilder.InterfaceBuilder.aJavaInterface;
import static com.mistraltech.smogen.codegenerator.javabuilder.InterfaceMethodBuilder.anInterfaceMethod;
import static com.mistraltech.smogen.codegenerator.javabuilder.MethodBuilder.aMethod;
import static com.mistraltech.smogen.codegenerator.javabuilder.MethodCallBuilder.aMethodCall;
import static com.mistraltech.smogen.codegenerator.javabuilder.NewInstanceBuilder.aNewInstance;
import static com.mistraltech.smogen.codegenerator.javabuilder.ParameterBuilder.aParameter;
import static com.mistraltech.smogen.codegenerator.javabuilder.ReturnStatementBuilder.aReturnStatement;
import static com.mistraltech.smogen.codegenerator.javabuilder.StaticMethodCallBuilder.aStaticMethodCall;
import static com.mistraltech.smogen.codegenerator.javabuilder.StaticVariableReaderBuilder.aStaticVariable;
import static com.mistraltech.smogen.codegenerator.javabuilder.TypeBuilder.aType;
import static com.mistraltech.smogen.codegenerator.javabuilder.TypeParameterBuilder.aTypeParameter;
import static com.mistraltech.smogen.codegenerator.javabuilder.TypeParameterDeclBuilder.aTypeParameterDecl;
import static com.mistraltech.smogen.utils.NameUtils.createFQN;

public class MatcherInterfaceCodeWriter extends AbstractMatcherCodeWriter {

    public MatcherInterfaceCodeWriter(MatcherGeneratorProperties matcherGeneratorProperties) {
        super(matcherGeneratorProperties);
    }

    @Override
    protected void generateDocumentContent(JavaDocumentBuilder document) {
        document.addInterface(generateMatcherInterface());
    }

    private InterfaceBuilder generateMatcherInterface() {
        final PsiPackage targetPackage = getPackage();
        final String generatedClassFQN = createFQN(targetPackage.getQualifiedName(), generatorProperties.getClassName());
        final TypeBuilder matcherType;

        InterfaceBuilder clazz = aJavaInterface()
                .withAccessModifier("public")
                .withName(generatorProperties.getClassName())
                .withTypeParameters(typeParameterDecls())
                .withAnnotation(anAnnotation()
                        .withType(aType()
                                .withName("com.mistraltech.smog.core.annotation.Matches"))
                        .withParameter(anExpression().withText("value=" + getSourceClassFQName() + ".class"))
                        .withParameter(anExpression().withText("description=\"" + getMatchedObjectDescription() + "\"")));

        final TypeBuilder matchedType = aType()
                .withName(getSourceClassFQName())
                .withTypeBindings(typeParameters());

        final TypeBuilder returnType;
        final TypeBuilder matchedTypeParam;

        if (generatorProperties.isExtensible()) {
            TypeParameterDeclBuilder returnTypeDecl = aTypeParameterDecl()
                    .withName("R")
                    .withExtends(aType()
                            .withName(generatedClassFQN)
                            .withTypeBindings(typeParameters())
                            .withTypeBinding(aTypeParameter()
                                    .withName("R"))
                            .withTypeBinding(aTypeParameter()
                                    .withName("T")));
            returnType = returnTypeDecl.getType();

            TypeParameterDeclBuilder matchedTypeDecl = aTypeParameterDecl().withName("T")
                    .withExtends(matchedType);

            matchedTypeParam = matchedTypeDecl.getType();

            matcherType = aType().withName(generatorProperties.getClassName())
                    .withTypeBinding(aTypeParameter())
                    .withTypeBinding(matchedType);

            clazz.withTypeParameter(returnTypeDecl)
                    .withTypeParameter(matchedTypeDecl);
        } else {
            returnType = aType()
                    .withName(generatedClassFQN)
                    .withTypeBindings(typeParameters());
            matchedTypeParam = matchedType;
            matcherType = returnType;
        }

        applySuperInterface(clazz, returnType, matchedTypeParam);

        applyInterfaceBody(clazz, returnType, matchedType, matcherType);

        return clazz;
    }

    private void applySuperInterface(InterfaceBuilder clazz, TypeBuilder returnType, TypeBuilder matchedTypeParam) {
        TypeBuilder superType;

        if (generatorProperties.getMatcherSuperClassName() != null) {
            superType = aType()
                    .withName(generatorProperties.getMatcherSuperClassName())
                    .withTypeBindings(getSourceSuperClassParameterBuilders())
                    .withTypeBinding(returnType)
                    .withTypeBinding(matchedTypeParam);
        } else {
            superType = aType()
                    .withName("org.hamcrest.Matcher")
                    .withTypeBinding(matchedTypeParam);
        }

        clazz.withImplementedInterface(superType);
    }

    private void applyInterfaceBody(InterfaceBuilder clazz, TypeBuilder returnType, TypeBuilder matchedType, TypeBuilder matcherType) {
        final boolean includeSuperClassProperties = generatorProperties.getMatcherSuperClassName() == null;
        final List<Property> sourceClassProperties = PropertyLocator.locateProperties(getSourceClass(), includeSuperClassProperties);

        if (generatorProperties.isGenerateFactoryMethods()) {
            clazz.withMethod(generateStaticFactoryMethod(matcherType));

            if (generatorProperties.isGenerateTemplateFactoryMethod()) {
                clazz.withMethod(generateLikeStaticFactoryMethod(matcherType, matchedType));
                clazz.withMethod(generateLikeMethod(returnType, matchedType));
            }
        }

        for (Property property : sourceClassProperties) {
            clazz.withMethods(generateMatcherSetters(property, returnType));
        }
    }

    private MethodSignatureBuilder generateStaticFactoryMethod(TypeBuilder matcherType) {
        final NewInstanceBuilder newInstance = aNewInstance()
                .withType(matcherType)
                .withParameter("MATCHED_OBJECT_DESCRIPTION");

        if (generatorProperties.isGenerateTemplateFactoryMethod()) {
            newInstance.withParameter("null");
        }

        MethodBuilder method = aMethod()
                .withStaticFlag(true)
                .withReturnType(matcherType)
                .withTypeParameters(typeParameterDecls())
                .withName(generatorProperties.getFactoryMethodPrefix() + getSourceClassName() + generatorProperties.getFactoryMethodSuffix())
                .withStatement(aReturnStatement()
                        .withExpression(aStaticMethodCall()
                                .withType(aType().withName("com.mistraltech.smog.proxy.javassist.JavassistMatcherGenerator"))
                                .withName("matcherOf")
                                .withParameter(aStaticVariable()
                                        .withType(matcherType)
                                        .withName("class"))));

        if (generatorProperties.isExtensible()) {
            method.withAnnotation(anAnnotation()
                    .withType(aType()
                            .withName("java.lang.SuppressWarnings"))
                    .withParameter(expressionText("\"unchecked\"")));
        }

        return method;
    }

    private MethodSignatureBuilder generateLikeStaticFactoryMethod(TypeBuilder matcherType, TypeBuilder matchedType) {
        MethodBuilder method = aMethod()
                .withStaticFlag(true)
                .withReturnType(matcherType)
                .withTypeParameters(typeParameterDecls())
                .withName(generatorProperties.getFactoryMethodPrefix() + getSourceClassName() + generatorProperties.getTemplateFactoryMethodSuffix())
                .withParameter(aParameter()
                        .withFinalFlag(generatorProperties.isMakeMethodParametersFinal())
                        .withType(matchedType)
                        .withName("template"))
                .withStatement(aReturnStatement()
                        .withExpression(aMethodCall()
                                .withObject(aStaticMethodCall()
                                        .withType(aType().withName("com.mistraltech.smog.proxy.javassist.JavassistMatcherGenerator"))
                                        .withName("matcherOf")
                                        .withParameter(aStaticVariable()
                                                .withType(matcherType)
                                                .withName("class")))
                                .withName("like")
                                .withParameter("template")));

        if (generatorProperties.isExtensible()) {
            method.withAnnotation(anAnnotation()
                    .withType(aType()
                            .withName("java.lang.SuppressWarnings"))
                    .withParameter(expressionText("\"unchecked\"")));
        }

        return method;
    }

    private MethodSignatureBuilder generateLikeMethod(TypeBuilder returnType, TypeBuilder matchedType) {
        return anInterfaceMethod()
                .withReturnType(returnType)
                .withName("like")
                .withParameter(aParameter()
                        .withFinalFlag(generatorProperties.isMakeMethodParametersFinal())
                        .withType(matchedType)
                        .withName("template"));
    }

    private List<InterfaceMethodBuilder> generateMatcherSetters(@NotNull Property property, TypeBuilder returnType) {
        final TypeBuilder boxedPropertyType = getPropertyTypeBuilder(property, true);

        final InterfaceMethodBuilder valueSetter = anInterfaceMethod()
                .withReturnType(returnType)
                .withName(setterMethodName(property))
                .withParameter(aParameter()
                        .withFinalFlag(generatorProperties.isMakeMethodParametersFinal())
                        .withType(getPropertyTypeBuilder(property, false))
                        .withName(property.getFieldName()));

        final InterfaceMethodBuilder matcherSetter = anInterfaceMethod()
                .withReturnType(returnType)
                .withName(setterMethodName(property))
                .withParameter(aParameter()
                        .withFinalFlag(generatorProperties.isMakeMethodParametersFinal())
                        .withType(aType()
                                .withName("org.hamcrest.Matcher")
                                .withTypeBinding(aTypeParameter()
                                        .withType(boxedPropertyType)
                                        .withSuperTypes(true)))
                        .withName(matcherAttributeName(property)));

        if (isCustomSetterName()) {
            final TypeBuilder matchesPropertyAnnotationType = aType().withName(MATCHES_PROPERTY_ANNOTATION_CLASS_NAME);
            final AnnotationBuilder matchesPropertyAnnotation = anAnnotation()
                    .withType(matchesPropertyAnnotationType)
                    .withParameter(anExpression()
                            .withText("\"" + property.getFieldName() + "\""));
            valueSetter.withAnnotation(matchesPropertyAnnotation);
            matcherSetter.withAnnotation(matchesPropertyAnnotation);
        }

        return Arrays.asList(valueSetter, matcherSetter);
    }

    private boolean isCustomSetterName() {
        return ! (generatorProperties.getSetterPrefix().equals(DEFAULT_SETTER_METHOD_PREFIX) &&
                generatorProperties.getSetterSuffix().equals(DEFAULT_SETTER_METHOD_SUFFIX));
    }
}

File: src/main/kotlin/com/mistraltech/smogen/codegenerator/matchergenerator/MatcherInterfaceCodeWriter.kt
package com.mistraltech.smogen.codegenerator.matchergenerator

import com.mistraltech.smogen.codegenerator.javabuilder.AnnotationBuilder
import com.mistraltech.smogen.codegenerator.javabuilder.ExpressionBuilder
import com.mistraltech.smogen.codegenerator.javabuilder.ExpressionTextBuilder
import com.mistraltech.smogen.codegenerator.javabuilder.InterfaceBuilder
import com.mistraltech.smogen.codegenerator.javabuilder.InterfaceMethodBuilder
import com.mistraltech.smogen.codegenerator.javabuilder.JavaDocumentBuilder
import com.mistraltech.smogen.codegenerator.javabuilder.MethodBuilder
import com.mistraltech.smogen.codegenerator.javabuilder.MethodCallBuilder
import com.mistraltech.smogen.codegenerator.javabuilder.MethodSignatureBuilder
import com.mistraltech.smogen.codegenerator.javabuilder.NewInstanceBuilder
import com.mistraltech.smogen.codegenerator.javabuilder.ParameterBuilder
import com.mistraltech.smogen.codegenerator.javabuilder.ReturnStatementBuilder
import com.mistraltech.smogen.codegenerator.javabuilder.StaticMethodCallBuilder
import com.mistraltech.smogen.codegenerator.javabuilder.StaticVariableReaderBuilder
import com.mistraltech.smogen.codegenerator.javabuilder.TypeBuilder
import com.mistraltech.smogen.codegenerator.javabuilder.TypeParameterBuilder
import com.mistraltech.smogen.codegenerator.javabuilder.TypeParameterDeclBuilder
import com.mistraltech.smogen.property.Property
import com.mistraltech.smogen.property.PropertyLocator.locatePropertiesFromGetters
import com.mistraltech.smogen.utils.NameUtils.createFQN

@Suppress("DuplicatedCode")
class MatcherInterfaceCodeWriter(matcherGeneratorProperties: MatcherGeneratorProperties) :
    AbstractMatcherCodeWriter(matcherGeneratorProperties) {

    override fun generateDocumentContent(document: JavaDocumentBuilder) {
        document.addInterface(generateMatcherInterface())
    }

    private fun generateMatcherInterface(): InterfaceBuilder {
        val generatedClassFQN = createFQN(targetPackage.qualifiedName, generatorProperties.className!!)

        val matcherType: TypeBuilder

        val clazz = InterfaceBuilder.aJavaInterface()
            .withAccessModifier("public")
            .withName(generatorProperties.className)
            .withTypeParameters(typeParameterDecls())
            .withAnnotation(
                AnnotationBuilder.anAnnotation()
                    .withType(
                        TypeBuilder.aType()
                            .withName("com.mistraltech.smog.core.annotation.Matches")
                    )
                    .withParameter(ExpressionBuilder.anExpression().withText("value=$sourceClassFQName.class"))
                    .withParameter(
                        ExpressionBuilder.anExpression().withText("description=\"$matchedObjectDescription\"")
                    )
            )

        val matchedType = TypeBuilder.aType()
            .withName(sourceClassFQName)
            .withTypeBindings(typeParameters())

        val returnType: TypeBuilder
        val matchedTypeParam: TypeBuilder

        if (generatorProperties.isExtensible) {
            val returnTypeDecl = TypeParameterDeclBuilder.aTypeParameterDecl()
                .withName("R")
                .withExtends(
                    TypeBuilder.aType()
                        .withName(generatedClassFQN)
                        .withTypeBindings(typeParameters())
                        .withTypeBinding(
                            TypeParameterBuilder.aTypeParameter()
                                .withName("R")
                        )
                        .withTypeBinding(
                            TypeParameterBuilder.aTypeParameter()
                                .withName("T")
                        )
                )

            returnType = returnTypeDecl.type

            val matchedTypeDecl = TypeParameterDeclBuilder.aTypeParameterDecl().withName("T")
                .withExtends(matchedType)

            matchedTypeParam = matchedTypeDecl.type

            matcherType = TypeBuilder.aType().withName(generatorProperties.className)
                .withTypeBinding(TypeParameterBuilder.aTypeParameter())
                .withTypeBinding(matchedType)

            clazz.withTypeParameter(returnTypeDecl)
                .withTypeParameter(matchedTypeDecl)
        } else {
            returnType = TypeBuilder.aType()
                .withName(generatedClassFQN)
                .withTypeBindings(typeParameters())
            matchedTypeParam = matchedType
            matcherType = returnType
        }

        applySuperInterface(clazz, returnType, matchedTypeParam)

        applyInterfaceBody(clazz, returnType, matchedType, matcherType)

        return clazz
    }

    private fun applySuperInterface(clazz: InterfaceBuilder, returnType: TypeBuilder, matchedTypeParam: TypeBuilder) {
        val superType: TypeBuilder = if (generatorProperties.matcherSuperClassName != null) {
            TypeBuilder.aType()
                .withName(generatorProperties.matcherSuperClassName)
                .withTypeBindings(sourceSuperClassParameterBuilders)
                .withTypeBinding(returnType)
                .withTypeBinding(matchedTypeParam)
        } else {
            TypeBuilder.aType()
                .withName("org.hamcrest.Matcher")
                .withTypeBinding(matchedTypeParam)
        }

        clazz.withImplementedInterface(superType)
    }

    private fun applyInterfaceBody(
        clazz: InterfaceBuilder,
        returnType: TypeBuilder,
        matchedType: TypeBuilder,
        matcherType: TypeBuilder
    ) {
        val includeSuperClassProperties = generatorProperties.matcherSuperClassName == null

        val sourceClassProperties = locatePropertiesFromGetters(sourceClass!!, includeSuperClassProperties)

        clazz.withMethod(generateStaticFactoryMethod(matcherType))

        if (generatorProperties.isGenerateTemplateFactoryMethod) {
            clazz.withMethod(generateLikeStaticFactoryMethod(matcherType, matchedType))
            clazz.withMethod(generateLikeMethod(returnType, matchedType))
        }

        sourceClassProperties.forEach { property ->
            clazz.withMethods(generateMatcherSetters(property, returnType))
        }
    }

    private fun generateStaticFactoryMethod(matcherType: TypeBuilder): MethodBuilder {
        val newInstance = NewInstanceBuilder.aNewInstance()
            .withType(matcherType)
            .withParameter("MATCHED_OBJECT_DESCRIPTION")

        if (generatorProperties.isGenerateTemplateFactoryMethod) {
            newInstance.withParameter("null")
        }

        val method = MethodBuilder.aMethod()
            .withStaticFlag(true)
            .withReturnType(matcherType)
            .withTypeParameters(typeParameterDecls())
            .withName(
                generatorProperties.factoryMethodPrefix + sourceClassName +
                    generatorProperties.factoryMethodSuffix
            )
            .withStatement(
                ReturnStatementBuilder.aReturnStatement()
                    .withExpression(
                        StaticMethodCallBuilder.aStaticMethodCall()
                            .withType(
                                TypeBuilder.aType()
                                    .withName("com.mistraltech.smog.proxy.javassist.JavassistMatcherGenerator")
                            )
                            .withName("matcherOf")
                            .withParameter(
                                StaticVariableReaderBuilder.aStaticVariable()
                                    .withType(matcherType)
                                    .withName("class")
                            )
                    )
            )

        if (generatorProperties.isExtensible) {
            method.withAnnotation(
                AnnotationBuilder.anAnnotation()
                    .withType(
                        TypeBuilder.aType()
                            .withName("java.lang.SuppressWarnings")
                    )
                    .withParameter(ExpressionTextBuilder.expressionText("\"unchecked\""))
            )
        }

        return method
    }

    private fun generateLikeStaticFactoryMethod(
        matcherType: TypeBuilder,
        matchedType: TypeBuilder
    ): MethodSignatureBuilder<*> {
        val method = MethodBuilder.aMethod()
            .withStaticFlag(true)
            .withReturnType(matcherType)
            .withTypeParameters(typeParameterDecls())
            .withName(
                generatorProperties.factoryMethodPrefix +
                    sourceClassName + generatorProperties.templateFactoryMethodSuffix
            )
            .withParameter(
                ParameterBuilder.aParameter()
                    .withFinalFlag(generatorProperties.isMakeMethodParametersFinal)
                    .withType(matchedType)
                    .withName("template")
            )
            .withStatement(
                ReturnStatementBuilder.aReturnStatement()
                    .withExpression(
                        MethodCallBuilder.aMethodCall()
                            .withObject(
                                StaticMethodCallBuilder.aStaticMethodCall()
                                    .withType(
                                        TypeBuilder.aType()
                                            .withName("com.mistraltech.smog.proxy.javassist.JavassistMatcherGenerator")
                                    )
                                    .withName("matcherOf")
                                    .withParameter(
                                        StaticVariableReaderBuilder.aStaticVariable()
                                            .withType(matcherType)
                                            .withName("class")
                                    )
                            )
                            .withName("like")
                            .withParameter("template")
                    )
            )

        if (generatorProperties.isExtensible) {
            method.withAnnotation(
                AnnotationBuilder.anAnnotation()
                    .withType(
                        TypeBuilder.aType()
                            .withName("java.lang.SuppressWarnings")
                    )
                    .withParameter(ExpressionTextBuilder.expressionText("\"unchecked\""))
            )
        }

        return method
    }

    private fun generateLikeMethod(returnType: TypeBuilder, matchedType: TypeBuilder): InterfaceMethodBuilder {
        return InterfaceMethodBuilder.anInterfaceMethod()
            .withReturnType(returnType)
            .withName("like")
            .withParameter(
                ParameterBuilder.aParameter()
                    .withType(matchedType)
                    .withName("template")
            )
    }

    private fun generateMatcherSetters(property: Property, returnType: TypeBuilder): List<InterfaceMethodBuilder> {
        val boxedPropertyType = getPropertyTypeBuilder(property, true)

        val valueSetter = InterfaceMethodBuilder.anInterfaceMethod()
            .withReturnType(returnType)
            .withName(setterMethodName(property))
            .withParameter(
                ParameterBuilder.aParameter()
                    .withType(getPropertyTypeBuilder(property, false))
                    .withName(property.fieldName)
            )

        val matcherSetter = InterfaceMethodBuilder.anInterfaceMethod()
            .withReturnType(returnType)
            .withName(setterMethodName(property))
            .withParameter(
                ParameterBuilder.aParameter()
                    .withType(
                        TypeBuilder.aType()
                            .withName("org.hamcrest.Matcher")
                            .withTypeBinding(
                                TypeParameterBuilder.aTypeParameter()
                                    .withType(boxedPropertyType)
                                    .withSuperTypes(true)
                            )
                    )
                    .withName(matcherAttributeName(property))
            )

        if (isCustomSetterName) {
            val matchesPropertyAnnotationType = TypeBuilder.aType()
                .withName(MATCHES_PROPERTY_ANNOTATION_CLASS_NAME)

            val matchesPropertyAnnotation = AnnotationBuilder.anAnnotation()
                .withType(matchesPropertyAnnotationType)
                .withParameter(
                    ExpressionBuilder.anExpression()
                        .withText("\"" + property.fieldName + "\"")
                )

            valueSetter.withAnnotation(matchesPropertyAnnotation)

            matcherSetter.withAnnotation(matchesPropertyAnnotation)
        }

        return listOf(valueSetter, matcherSetter)
    }

    private val isCustomSetterName: Boolean
        get() = generatorProperties.setterPrefix != DEFAULT_SETTER_METHOD_PREFIX ||
            generatorProperties.setterSuffix != DEFAULT_SETTER_METHOD_SUFFIX
}
